frame/composable-traits/src/undercollateralized_loans/configs.rs

```diff
@@ -0,0 +1,299 @@
+	BlockNumber: Clone + Eq + PartialEq,
+	VaultId: Clone + Eq + PartialEq,
+{
+	pub fn new(
+		account_id: AccountId,
+		manager: AccountId,
+		borrow_asset_vault: VaultId,
+		borrow_asset_id: AssetId,
+		collateral_asset_id: AssetId,
+		max_price_age: BlockNumber,
+		whitelist: BTreeSet<AccountId>,
+	) -> Self {
+		Self {
+			account_id,
+			manager,
+			borrow_asset_vault,
+			borrow_asset_id,
+			collateral_asset_id,
+			max_price_age,
+			whitelist,
+		}
+	}
+
+	/// Get a reference to the market config's account id.
+	pub fn account_id(&self) -> &AccountId {
+		&self.account_id
+	}
+
+	/// Get a reference to the market config's manager.
+	pub fn manager(&self) -> &AccountId {
+		&self.manager
+	}
+	/// Get a reference to the market config's borrow asset vault.
+	pub fn borrow_asset_vault(&self) -> &VaultId {
+		&self.borrow_asset_vault
+	}
+
+	/// Get a reference to the market config's borrow asset.
+	pub fn borrow_asset_id(&self) -> &AssetId {
+		&self.borrow_asset_id
+	}
+
+	/// Get a reference to the market config's collateral asset.
+	pub fn collateral_asset_id(&self) -> &AssetId {
+		&self.collateral_asset_id
+	}
+
+	/// Get a reference to the market config's max price age.
+	pub fn max_price_age(&self) -> &BlockNumber {
+		&self.max_price_age
+	}
+
+	/// Get a reference to the market config's whitelist.
+	#[must_use]
+	pub fn whitelist(&self) -> &BTreeSet<AccountId> {
+		&self.whitelist
+	}
+}
+
+#[derive(Encode, Decode, Default, TypeInfo, RuntimeDebug, Clone, Eq, PartialEq)]
+pub struct LoanConfig<AccountId, AssetId, Balance, Timestamp>
+where
+	AccountId: Clone + Eq + PartialEq,
+	AssetId: Clone + Eq + PartialEq,
+	Balance: Clone + Eq + PartialEq,
+	Timestamp: Clone + Eq + PartialEq,
+{
+	/// Loan account id.
+	account_id: AccountId,
+	/// Market account id.
+	market_account_id: AccountId,
+	/// Borrower account id.
+	/// Should be whitelisted.
+	borrower_account_id: AccountId,
+	/// The asset being used as collateral.
+	collateral_asset_id: AssetId,
+	/// The asset being used as borrow asset.
+	borrow_asset_id: AssetId,
+	/// Amount of borrowed money.  
+	principal: Balance,
+	/// Amount of assets which should be putted as collateral.
+	collateral: Balance,
+	/// Schedule of payments
+	schedule: BTreeMap<Timestamp, Balance>,
+	/// The moment of the first interest payment.
+	first_payment_moment: Timestamp,
+	/// The moment of the last interest payment and principal repayment.
+	last_payment_moment: Timestamp,
+}
+
+impl<AccountId, AssetId, Balance, Timestamp> LoanConfig<AccountId, AssetId, Balance, Timestamp>
+where
+	AccountId: Clone + Eq + PartialEq,
+	AssetId: Clone + Eq + PartialEq,
+	Balance: Clone + Eq + PartialEq,
+	Timestamp: Clone + Eq + PartialEq + Ord,
+{
+	pub fn new(
+		account_id: AccountId,
+		market_account_id: AccountId,
+		borrower_account_id: AccountId,
+		collateral_asset_id: AssetId,
+		borrow_asset_id: AssetId,
+		principal: Balance,
+		collateral: Balance,
+		schedule: Vec<(Timestamp, Balance)>,
```

Comment:
Why not take the map as input? BTrees are sorted, so you can `.first()`/`.last()` on them

Reply:
@benluelo first_entry() method is nightly-only https://paritytech.github.io/substrate/master/sp_std/collections/btree_map/struct.BTreeMap.html#method.first_entry

Reply:
Hmm, I didn't realize it was still on nightly. It's been stuck there for quite a while it seems

Reply:
Either way, you can just take the map directly as input instead of collecting into it immediately